core/src/main/scala/kafka/server/ReplicationQuotaManager.scala

```diff
@@ -0,0 +1,192 @@
+  * limitations under the License.
+  */
+package kafka.server
+
+import java.util.concurrent.{ConcurrentHashMap, TimeUnit}
+import kafka.common.TopicAndPartition
+import kafka.server.Constants._
+import kafka.server.ReplicationQuotaManagerConfig._
+import kafka.utils.CoreUtils._
+import kafka.utils.Logging
+import org.apache.kafka.common.MetricName
+import org.apache.kafka.common.metrics._
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import org.apache.kafka.common.metrics.stats.{Rate, SimpleRate}
+import org.apache.kafka.common.utils.Time
+
+/**
+  * Configuration settings for quota management
+  *
+  * @param quotaBytesPerSecondDefault The default bytes per second quota allocated to internal replication
+  * @param numQuotaSamples            The number of samples to retain in memory
+  * @param quotaWindowSizeSeconds     The time span of each sample
+  *
+  */
+case class ReplicationQuotaManagerConfig(quotaBytesPerSecondDefault: Long = QuotaBytesPerSecondDefault,
+                                         numQuotaSamples: Int = DefaultNumQuotaSamples,
+                                         quotaWindowSizeSeconds: Int = DefaultQuotaWindowSizeSeconds)
+
+object ReplicationQuotaManagerConfig {
+  val QuotaBytesPerSecondDefault = Long.MaxValue
+  // Always have 10 whole windows + 1 current window
+  val DefaultNumQuotaSamples = 11
+  val DefaultQuotaWindowSizeSeconds = 1
+  // Purge sensors after 1 hour of inactivity
+  val InactiveSensorExpirationTimeSeconds = 3600
+}
+
+trait ReplicaQuota {
+  def isThrottled(topicAndPartition: TopicAndPartition): Boolean
+  def isQuotaExceeded(): Boolean
+}
+
+object Constants {
+  val allReplicas = Seq[Int](-1)
+}
+
+/**
+  * Tracks replication metrics and comparing them to any quotas for throttled partitions.
+  *
+  * @param config          The quota configs
+  * @param metrics         The Metrics instance
+  * @param replicationType The name / key for this quota manager, typically leader or follower
+  * @param time            Time object to use
+  */
+class ReplicationQuotaManager(val config: ReplicationQuotaManagerConfig,
+                              private val metrics: Metrics,
+                              private val replicationType: String,
+                              private val time: Time) extends Logging with ReplicaQuota {
+  private val lock = new ReentrantReadWriteLock()
+  private val throttledPartitions = new ConcurrentHashMap[String, Seq[Int]]()
+  private var quota: Quota = null
+  private val sensorAccess = new SensorAccess
+  private val rateMetricName = metrics.metricName("byte-rate", replicationType, s"Tracking byte-rate for $replicationType")
+
+  /**
+    * Update the quota
+    *
+    * @param quota
+    */
+  def updateQuota(quota: Quota) = {
+    inWriteLock(lock) {
+      this.quota = quota
+      if (metrics.metrics.containsKey(rateMetricName))
+        metrics.metrics.get(rateMetricName).config(getQuotaMetricConfig(quota))
```

Comment:
The above may not be 100% safe since the metric could be expired and removed between the two statements. It's safer if we save metrics.metrics.get(rateMetricName) to a local val and then check null and update the config.